sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+            TRUE,
+            NORMAL_PRIORITY_CLASS | CREATE_NO_WINDOW,
+            lpEnvironment,
+            NULL,
+            &startupInfo,
+            &procInfo),
+        "Cannot create process");
+  }
+
+  // We won't be needing the process main thread handle on our end.
+  static_cast<void>(CloseHandle(procInfo.hThread));
+
+  // Keep the process handle so we can cancel it if it takes too long.
+  m_processHandle = procInfo.hProcess;
+
+  // We won't be writing to the pipe that is meant for the process.
+  // We will only be reading the pipe.
+  // So, now that the process is started, we can close write handle on our end.
+  static_cast<void>(CloseHandle(outputPipe.m_writeHandle));
+  outputPipe.m_writeHandle = NULL;
+#else
+  // Form the 'argv' array:
+  // * An array of pointers to non-const C strings (0-terminated).
+  // * Last element is NULL pointer.
+  // * First element (at index 0) is path to a program.
+  {
+    // Since the strings that argv is pointing at do need to be non-const,
+    // and also because each commnd line argument needs to be a separate 0-terminated string,
+    // We do form all their values in the m_argvValues.
+
+    // Since we append m_argvValues as we go, at one point after insertion it may reallocate the
+    // buffer to a different address in memory. For that reason, we can't grab addresses before we
+    // are done forming m_argvValues contents - so until that we record indices where each string
+    // start - in argvValuePositions.
+    {
+      std::vector<decltype(m_argvValues)::size_type> argvValuePositions;
+
+      // First string is the path to executable, and not the actual first argument.
+      {
+        std::string const Shell = "/bin/sh";
+        EnsureShellExists(Shell);
+        AppendToArgvValues(m_argvValues, argvValuePositions, Shell);
+      }
+
+      // Second argument is the shell switch that tells the command line shell to execute a command
+      AppendToArgvValues(m_argvValues, argvValuePositions, "-c");
+
+      // Third value is the command that needs to be executed.
+      AppendToArgvValues(m_argvValues, argvValuePositions, command);
+
+      // We are done appending to m_argvValues, so it is now safe to grab addresses to the elements
+      // in it.
+      for (auto const pos : argvValuePositions)
+      {
+        m_argv.push_back(m_argvValues.data() + pos);
+      }
+    }
+
+    // argv last element needs to be NULL pointer.
+    m_argv.push_back(NULL);
+  }
+
+  // Form the 'envp' array:
+  // * An array of pointers to non-const C strings (0-terminated).
+  // * Strings are in form key=value (PATH uses ':' as separator)
+  // * Last element is NULL pointer.
+  // * First element (at index 0) is path to a program.
+  {
+    auto const actualPathVarValue = Environment::GetVariable("PATH");
+    auto const processPathVarStatement = std::string("PATH=") + actualPathVarValue
+        + (actualPathVarValue.empty() ? "" : ":") + "/usr/bin:/usr/local/bin";
```

Comment:
You're assuming that az is in /usr/bin or /usr/local/bin without actually checking to see if az is in those directories. If you don't trust that the OS has put cmd.exe/sh in the path, why are you then trusting that az/az.cmd are in their expected location? 

IMHO you check every command you're going to launch or you check none of them.

Reply:
We just append these directories (typical az cli install locations) to the end of the `PATH` variable (other SDKs also do this).

For the cmd.exe, we need full path to executable in order to start the process. Then, we rely on the cmd/sh to be able to find `az` via `PATH` variable. We really don't need to know the path. If we knew, I'd rather start `az` process directly, without starting `cmd` or `sh`. I wish we didn't need to know the full path for the executable, but API requires it.
Especially it sucks on POSIX: if you give `posix_spawn()` a path to a non-existent binary, it would not fail. But instead you get a zombie process, and then you need to do some tricks in order to discover that it is in fact a zombie.

Reply:
You need to run the shell because az is a batch file on Windows and Linux, so it has to be run by a script interpreter (btw, az is a bash script on linux so running /bin/sh is inefficient since it immediately turns around and invokes bash).

Reply:
You can't successfully invoke `CreateProcessA()` with `cmd.exe` either as first or the second parameter.

Reply:
Fascinating. The [CreateProcess documentation](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessa) implies differently:
```
To run a batch file, you must start the command interpreter; set lpApplicationName to cmd.exe and set lpCommandLine to the following arguments: /c plus the name of the batch file.
```

But you've tried it in the wild and obviously the documentation is incorrect.

Reply:
I set `lpApplicationName` to `cmd.exe`, and put the rest into `lpCommandLine`, it is `/c az account ...` exactly as described in the documentation, but `CreateProcessA()` returns `FALSE` and `GetLastError()` is `ERROR_FILE_NOT_FOUND`.

Reply:
Grr.. What about cmd /c az account?

That's indescribably annoying, TBH.

Reply:
That one did work! So, no full path is needed. I updated the code.